### PR TITLE
[FEEDBACK WANTED] Prefs UI Rework! [IDB IGNORE]

### PR DESCRIPTION
**Pull request by RimiNosha**
# Want to provide feedback? Comment here, or PING me on discord.
See proof of testing to see what I've massacred so far.

## About The Pull Request

What a way to return to trying to play the game, huh? I tried to make a character and forgot how good I had it when I was doing my own thing, so I got so mad that I got maint approval to rip apart the current UI.

So, enjoy parts of what I learned on another codebase, except nearly entirely rewritten to be """modular friendly""".

Oh, the code is also entirely unfinished, and I'm doing a lot of lazy naming/copy-pasting to get an initial concept out. There's no point in entirely refining the thing if it turns out folk don't actually want this.

I also have done zero self-review, so there's probably some hideous code and a lot of missing comments.

## Why It's Good For The Game

Shit man, it's been a constant complaint since tgui prefs first dropped. The current menu is a massive fuck off scroll list that necessitates multiple master toggles lest the average player be bombarded with over 170 prefs at once.

So yeah, making an actual player creator menu system is really good for the game.

### Notable Changes:
- The prefs menu is now 1200px wide instead of 920px to give a little more breathing space. Most notable with the jobs menu. It can be put back into being fullscreen if it's really wanted, though.
- The prefs menu now loads into an index page with a large amount of categories for things that would have previously been tabs.
- To alleviate the lag caused by remounting byondui components, I've made the character preview a static thing. This should improve lag spikes when clicking through preferences measurably.
- Many, many prefs have proper icon generation (WIP) and have proper previews at long last.
- Master pref toggles are entirely gone.
  - Instead these are housed in the new:
- **Major Preference Details**
  - Any major preference that also has an icon will get ALL related preferences put into this new details pane.
- Searching through icon prefs sucks DRASTICALLY less than previously. No more scrolling through what feels like an infinite list on a TINY popup.
  - The search bar inside said icon prefs also no longer scrolls with it's contents, and sticks with the header.
- Job prefs have been slightly reorganised to better fit on the screen.
  - Isolated enough to be it's own PR without conflicts if maints desire.
- Food prefs are no longer it's own window, and is now a character preference category.
- Species no longer sets the player species on click, you now have to press confirm for anything to happen to your beloved character.
  - Misc formatting changes also took place here.

### Notable Code Changes:
- There are now a scattershot of bubber-specific pref pages now, and some normal preference elements and functions have been entirely supplanted by their modular versions.
  - **Double check what version of a function or element you're importing is before using them.**
- A LOT of skyrat-era modular edits have been reduced or entirely scrapped.
- Most preference categories are now under a bubber-specific define.

## Proof Of Testing
<details>
<summary>Screenshots/Videos</summary>

Index page so far (heavy WIP)
image

New player popup
image

What it looks like inside a category (for text-heavy ones, the bottom half takes the whole screen, and there's no icon pane (WIP))
image

The new details window
image

Oh yeah, the searchbar also isn't in the search area anymore
image

New species page
image

</details>

## Changelog
:cl:
qol: Overhaul the prefs UI to be much better for the prefs density on the server!
/:cl:

### CODE REVIEW

**Comment by ReturnToZender** [comment]
Rimi please you can't just bring this kinda heat day one holy shit (positive)

**Comment by RimiNosha** [comment]
Notes while walking to work and having slept on it:

Hair gradients really should still be kept under hair prefs with the old popout menu.

See if the category pages can be laid out similar to the details pages.

Icons for the category buttons? Might make it too busy, though.


**Later PRs**:
Add a tag system for hairstyles. There's so goddamn many that it'll help folk looking for a particular type in a very noticeable way.

This will sadly require me to dig through every single hairstyle.

**Comment by Strazyplus** [comment]
**oh god, oh god... i'm gonna...**  

C-can we get a pop out preview? so if our characters are large or taur. we can scale the preview and see them fully?

**Comment by RimiNosha** [comment]
No promises, but I can try

**Comment by Strazyplus** [comment]
> No promises, but I can try

Fun fact once upon a time there was a pop out window for preview that you could resize to see your character completely.

**Comment by RimiNosha** [comment]
Yeah, this looks a lot better
image

**Comment by ReturnToZender** [comment]
I think this has a lot of potential. Not a huge fan of the way the Index is laid out, I actually really like the buttons at the top of the screen, but the individual pages are awesome.

I'd _really_ like to see horns, ears, etc. finally moved to click-through menus instead of dropdowns. And eye customization so I can have functioning eyes!

**Comment by RimiNosha** [comment]
> I think this has a lot of potential. Not a huge fan of the way the Index is laid out

Yeah, I've not really spent any real time on the index yet. 

> I'd _really_ like to see horns, ears, etc. finally moved to click-through menus instead of dropdowns.

Already being implemented.

> And eye customization so I can have functioning eyes!

In a separate PR? Sure. I'd avoid adding new prefs in this PR though.

**Comment by ReturnToZender** [comment]
I'll be watching! I'm pretty stoked to have a cleanup pass in the character creator, it'll really put us ahead and help people make characters easier.

**Comment by RimiNosha** [comment]
TODO: Write unit test to scream about any invalid supplementary prefs, fix erroneous icon cuts, look through every icon and see if anything drastically broke.

Also figure out why the fuck none prefs are causing prefs assets to be yeeted into the sun for no apparent reason.

Also probably get a hold of a player save (or two) with permission so I can make sure I didn't colossally fuck up prefs.

**Comment by Iamgoofball** [comment]
I'm sorry but this looks absolutely awful. This is a massive, massive downgrade and shoves a bunch of things into submenus that didn't need to be in submenus. Please take a UX course.

**Comment by RimiNosha** [comment]
TL;DR from discord:

I agree that the UI is still pretty crap in it's current form. That's why it's a draft.

The submenu issue is a valid issue, and I've an idea on how to make it less awful, but honestly that's best left to a different PR.

Most of the finalising of the UI side of things won't be happening until the data the UI gets isn't garbage, so that's why it's an unholy amalgamation of bad choices.

**Comment by RimiNosha** [comment]
I've realised that my patience with TGUI is wearing thinner than I thought it would, so I'm going to tap out and just add another row of featureprefs to the old UI and call it a day. 

Closing this PR in the meantime as there's nothing new to show anymore, and it's commit history is a mess, and was never intended to be merged anyways.

The DM-side changes will still be happening, though, considering I need those damned icons, and I'm not working with my old flawed icon generation code.
